binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/InteractiveQueryService.java

```diff
@@ -105,32 +100,34 @@ public <T> T getQueryableStore(String storeName, QueryableStoreType<T> storeType
 			if (store != null) {
 				return store;
 			}
-			else if (contextSpecificKafkaStreams != null) {
-				LOG.warn("Store " + storeName
-						+ " could not be found in Streams context, falling back to all known Streams instances");
+			if (contextSpecificKafkaStreams != null) {
+				LOG.warn("Store (" + storeName + ") could not be found in Streams context, falling back to all known Streams instances");
 			}
-			final Set<KafkaStreams> kafkaStreams = kafkaStreamsRegistry.getKafkaStreams();
-			final Iterator<KafkaStreams> iterator = kafkaStreams.iterator();
-			while (iterator.hasNext()) {
+
+			for (KafkaStreams kafkaStreamApp : kafkaStreamsRegistry.getKafkaStreams()) {
 				try {
-					store = iterator.next()
-							.store(StoreQueryParameters.fromNameAndType(
-									storeName, storeType));
+					return getStateStoreFromKafkaStreams(kafkaStreamApp, storeName, storeType);
 				}
-				catch (InvalidStateStoreException e) {
-					// pass through..
-					throwable = e;
+				catch (Exception ex) {
+					throwable = ex;
 				}
 			}
-			if (store != null) {
-				return store;
-			}
-			throw new IllegalStateException(
-					"Error when retrieving state store: " + storeName,
-					throwable);
+			throw new IllegalStateException("Error retrieving state store: " + storeName, throwable);
 		});
 	}
 
+	private <T> T getStateStoreFromKafkaStreams(KafkaStreams kafkaStreams, String storeName, QueryableStoreType<T> storeType) {
+		// Check KafkaStreams app knows about the state store
+		T store = kafkaStreams.store(StoreQueryParameters.fromNameAndType(storeName, storeType));
+
+		// Check KafkaStreams app actually has the state store
```

Comment:
This is the extra check that fixes the issue. If you comment out these lines you will see the newly added tests fail.